### ISSUE TITLE
Issue with _StopAsync()

### ISSUE THREAD

**Issue by Camel-RD** [dranger003/llama.cpp-dotnet#11]
Program often gets stuck on this line
`await (_mainLoop ?? Task.CompletedTask);`
in _StopAsync()
One possible solution for this issue could be to rewrite this line like this:
`await (_mainLoop ?? Task.CompletedTask).ConfigureAwait(false);`

**Comment by dranger003** [comment]
Can you give more details how to reproduce this? I didn't experience it. If the main loop isn't returning then it must be awaiting on something and that would be causing a deadlock.

I saw they added a cancel callback in the upstream repo so I might be updating the code to avoid getting locked in `llama_decode`.

**Comment by Camel-RD** [comment]
I'm not exactly sure why, but this code will lock a Windows Forms application:
```
        async Task TestAwait()
        {
            var task = Task.Run(() =>
            {
                Task.Delay(3000).Wait();
            });
            await task;
        }

        private async void button1_Click(object sender, EventArgs e)
        {
            TestAwait().Wait();
            MessageBox.Show("Done");
        }

```
And LlmEngine.Dispose() has similar code.

**Comment by dranger003** [comment]
Try this:
```
    public partial class Form1 : Form
    {
        public Form1()
        {
            InitializeComponent();
        }

        private Task? _task = null;

        private void button1_Click(object sender, EventArgs e)
        {
            _task = Task.Run(() =>
            {
                Debug.WriteLine($"-> button1_Click()");
                Task.Delay(5000).Wait();
                Debug.WriteLine($"<- button1_Click()");
            });
        }

        private async void button2_Click(object sender, EventArgs e)
        {
            Debug.WriteLine($"-> button2_Click()");
            await (_task ?? Task.CompletedTask);
            Debug.WriteLine($"<- button2_Click()");
        }
    }
```

**Comment by Camel-RD** [comment]
I know this will work. But with task.Wait(), like its done in that Dispose() method, it will lock the application.

**Comment by dranger003** [comment]
EDIT: I added the `ConfigureAwait()` just now.

What if I add an async version of unload, you could then await it before quitting your app?
```
        public async Task UnloadModelAsync()
        {
            await _StopAsync();

            _batch.Dispose();
            _context.Dispose();
            _model.Dispose();
        }
```

**Comment by Camel-RD** [comment]
This .ConfigureAwait(false) should solve the issue. I think it's important to fix Dispose() method to be sure that using(var llmengine = new LlmEngine()){} is working properly.

**Comment by dranger003** [comment]
Fixed in commit ec27b99.
